pkg/operator/staticpod/certsyncpod/certsync_controller.go

```diff
@@ -257,29 +239,103 @@ func (c *CertSyncController) sync(ctx context.Context, syncCtx factory.SyncConte
+		eventRecorder.Warningf("CertificateUpdateFailed",
+			"Failed creating content directory for %s: %s/%s: %v", kind, objectNamespace, objectName, err)
+		errors = append(errors, err)
+		return errors
+	}
+
+	// Create a tmp source directory to be swapped.
+	srcDir, err := fs.MkdirTemp(filepath.Dir(contentDir), filepath.Base(contentDir)+"-*")
+	if err != nil {
+		eventRecorder.Warningf("CertificateUpdateFailed",
+			"Failed to create source %s directory for %s/%s: %v", kind, objectNamespace, objectName, err)
+		errors = append(errors, err)
+		return errors
+	}
+	defer fs.RemoveAll(srcDir)
+
+	// Populate the tmp directory with files.
+	for filename, content := range files {
+		fullFilename := filepath.Join(srcDir, filename)
+		klog.Infof("Writing %s manifest %q ...", kind, fullFilename)
+
+		if err := fs.WriteFile(fullFilename, []byte(content), filePerm); err != nil {
+			eventRecorder.Warningf("CertificateUpdateFailed",
+				"Failed writing file for %s: %s/%s: %v", kind, objectNamespace, objectName, err)
 			errors = append(errors, err)
 			continue
 		}
-		for filename, content := range secret.Data {
-			// TODO fix permissions
-			fullFilename := filepath.Join(contentDir, filename)
-			// if the existing is the same, do nothing
-			if reflect.DeepEqual(data[fullFilename], content) {
-				continue
-			}
+	}
+	if len(errors) > 0 {
+		return errors
+	}
 
-			klog.Infof("Writing secret manifest %q ...", fullFilename)
-			if err := staticpod.WriteFileAtomic(content, 0600, fullFilename); err != nil {
-				c.eventRecorder.Warningf("CertificateUpdateFailed", "Failed writing file for secret: %s/%s: %v", secret.Namespace, secret.Name, err)
-				errors = append(errors, err)
-				continue
-			}
+	// Make sure the tmp directory is correct regarding the target state.
+	for filename := range files {
+		fullFilename := filepath.Join(srcDir, filename)
+		content, err := fs.ReadFile(fullFilename)
+		if err != nil {
+			eventRecorder.Warningf("CertificateUpdateFailed",
+				"Failed to check %s file for %s/%s is correctly written: %v", kind, objectNamespace, objectName, err)
+			errors = append(errors, err)
+			continue
+		}
+		if !bytes.Equal(content, []byte(files[filename])) {
+			eventRecorder.Warningf("CertificateUpdateFailed",
+				"Content mismatch detected for temporary %s file for %v/%v", kind, objectNamespace, objectName)
+			errors = append(errors, fmt.Errorf("checksum mismatch detected for %q", fullFilename))
+			continue
 		}
-		c.eventRecorder.Eventf("CertificateUpdated", "Wrote updated secret: %s/%s", secret.Namespace, secret.Name)
+	}
+	if len(errors) > 0 {
+		return errors
 	}
 
-	return utilerrors.NewAggregate(errors)
+	// Swap directories atomically.
+	if err := fs.SwapDirectoriesAtomic(contentDir, srcDir); err != nil {
+		eventRecorder.Warningf("CertificateUpdateFailed",
+			"Failed to enable new %s directory for %s/%s: %v", kind, objectNamespace, objectName, err)
+		errors = append(errors, err)
```

Comment:
Restore previous dir contents here

Reply:
There is nothing to restore in the current logic, because we write the whole secrets/configmap directory anew and then we swap it atomically. The current state is not modified until it's finally swapped.

Reply:
So naturally when there is an error preparing that new copy, which is called `srcDir` here, we can just delete it and the current state is unaffected.